backend/src/git.rs

```diff
@@ -394,6 +616,37 @@ impl Interface {
         Ok(())
     }
 
+    /// A code level re-implementation of `git merge` for a specified remote branch.
+    /// This function handles the merge process for pulling changes specifically from the given branch.
+    fn git_merge_from_branch(
+        repo: &Repository,
+        remote_branch: &str,
+        fetch_commit: AnnotatedCommit<'_>,
+    ) -> Result<()> {
+        // First perform a merge analysis to understand how to proceed
+        let analysis = repo.merge_analysis(&[&fetch_commit])?;
+
+        // Handle fast-forward merges
+        if analysis.0.is_fast_forward() {
+            debug!("Performing fast forward merge from branch '{}'", remote_branch);
+            let refname = format!("refs/heads/{}", remote_branch);
+            let mut reference = repo.find_reference(&refname)?;
+            Self::fast_forward(repo, &mut reference, &fetch_commit)?;
+        } 
+        // Handle normal merges
+        else if analysis.0.is_normal() {
+            debug!("Performing normal merge from branch '{}'", remote_branch);
+            let head_commit = repo.reference_to_annotated_commit(&repo.head()?)?;
+            Self::normal_merge(repo, &fetch_commit, &head_commit)?;
+        } 
+        // If no merging is needed
+        else {
+            debug!("No work needed to merge from branch '{}'", remote_branch);
+        }
+
+        Ok(())
+    }
+
```

Comment:
Just have merge from branch replace `git_merge` entirely, it's a more flushed out version of the same function